bam2depth.c

```diff
@@ -41,377 +43,883 @@ DEALINGS IN THE SOFTWARE.  */
+
+    if (dh->end >= 0 && end_pos > dh->end)
+        end_pos = dh->end;
+    if (dh->end_pos[file] < end_pos)
+        dh->end_pos[file] = end_pos;
+
+    return 0;
+}
+
+// Hash on name -> alignment end pos. This permits a naive overlap removal.
+// Note it cannot analyse the overlapping sequence and qualities, so the
+// interaction of basecalls/qualities and the -Q parameter cannot be
+// applied here (unlike the full mpileup algorithm).
+KHASH_MAP_INIT_STR(olap_hash, hts_pos_t)
+typedef khash_t(olap_hash) olap_hash_t;
+
+static int fastdepth_core(depth_opt *opt, uint32_t nfiles, char **fn,
+                          samFile **fp, hts_itr_t **itr, sam_hdr_t **h) {
+    int ret = -1, err = 1, i;
+    olap_hash_t **overlaps = NULL;
+    depth_hist dh = {0};
+
+    // An array of bam structs, one per input file, to hold the next entry
+    bam1_t **b = calloc(nfiles, sizeof(*b));
+    int *finished = calloc(nfiles, sizeof(*finished)), to_go = nfiles;
+    if (!b || !finished)
+        goto err;
+
+    for (i = 0; i < nfiles; i++)
+        if (!(b[i] = bam_init1()))
+            goto err;
+
+    // Do we need one overlap hash per file? Or shared?
+    if (opt->remove_overlaps) {
+        if (!(overlaps = calloc(nfiles, sizeof(*overlaps))))
+            return -1;
+        for (i = 0; i < nfiles; i++) {
+            if (!(overlaps[i] = kh_init(olap_hash)))
+                return -1;
         }
-        if (hts_set_opt(data[i]->fp, CRAM_OPT_DECODE_MD, 0)) {
-            print_error_errno("depth", "Failed to set CRAM_OPT_DECODE_MD value");
-            status = EXIT_FAILURE;
-            goto depth_end;
+    }
+
+    // Create the initial histogram
+    dh.nfiles = nfiles;
+    dh.size = 0;
+    dh.hist = NULL;
+    dh.last_ref = -99;
+    dh.end_pos = NULL;
+    dh.last_output = itr && itr[0] ? itr[0]->beg : 0;
+    ks_initialize(&dh.ks);
+
+    // Clip results to region if specified
+    dh.beg = -1;
+    dh.end = -1;
+    dh.tid = 0;
+    if (itr && itr[0]) {
+        dh.tid = itr[0]->tid;
+        dh.beg = itr[0]->beg;
+        dh.end = itr[0]->end;
+    }
+
+    if (opt->header) {
+        fprintf(opt->out, "#CHROM\tPOS");
+        for (i = 0; i < nfiles; i++)
+            fprintf(opt->out, "\t%s", fn[i]);
+        fputc('\n', opt->out);
+    }
+
+    // Populate first record per file
+    for (i = 0; i < nfiles; i++) {
+        for(;;) {
+            ret = itr && itr[i]
+                ? sam_itr_next(fp[i], itr[i], b[i])
+                : sam_read1(fp[i], h[i], b[i]);
+            if (ret < -1)
+                goto err;
+            if (ret == -1) {
+                to_go--;
+                finished[i] = 1;
+                break;
+            }
+
+            if (b[i]->core.tid < 0)
+                continue;
+            if (b[i]->core.flag & opt->flag)
+                continue;
+            if (b[i]->core.qual < opt->min_mqual)
+                continue;
+
+            // FIXME?  Should this be sequence length minus soft-clips?
```

Comment:
Since the read start is from the soft clipped position there is a good case to be made for using the clipped length.  I guess the original depth did not do this though.

Reply:
I'd forgotten about that.  Yes it probably needs improving with an aligned-length function.  I did later realise why it calls the function to compute the sequence length from CIGAR instead of just using `b->core.l_qseq`, which totally baffled me originally.  It's because sequence `*` has length 0, but the min length parameter still wishes to accept it if CIGAR indicates it spans a large region.

Yes my intention was to have this in two stages.  The first is to get as bug-compatible as possible with the former samtools depth so it's easy to validate the impact of the rewrite, followed up with fixing what I view as bugs in the original implementation.

I did originally have it split up, but I think I got overly tidy and squashed it together, losing this separation.  Hence the bug fix for PROPER_PAIR vs PAIRED flag is now in there already IIRC.  Or maybe I figured it's only for overlap removal which is already different so not possible to validate anyway.  Not sure now.